Performance dashboard

avinash403 opened the pull request:
fixes #990

NOTE for testers: test it with `billing.faveohelpdesk.com`'s database to see a significant difference

ashu555 commented on app/Http/Controllers/DashboardController.php:
```diff
@@ -6,10 +6,13 @@
 use Illuminate\Http\Request;
+use phpDocumentor\Reflection\Types\Collection;
 
 class DashboardController extends Controller
```
Can you check Create new User button url on the dashboard by clicking on it.

ashu555 replied:
Also if the admin logs in, do we need to change the url to /dashboard?

avinash403 replied:
> Also if the admin logs in, do we need to change the url to /dashboard?

laravel files has been changed. We will do it when we correct those

avinash403 replied:
> Please check Create new User button url on the dashboard by clicking on it.

fixed

avinash403 commented:
> @ashu555 Please can you share us billing database so that I can performance for the dashboard.
> 
> Please send it on my mail [sarfraz.ansari@ladybirdweb.com](mailto:sarfraz.ansari@ladybirdweb.com).

sent via skype

ashu555 commented:
> first I checked load time with the development branch and then checked load time with your branch
> with billing DB. All data like total sales in INR and USD, Yearly , Monthly, Pending, and total users count showing correctly. If I miss anything please let me I will check.
> Please see the below docs.
> https://docs.google.com/document/d/10rh-LPE3HegoSMf5rAO1GyYzxCLRu04vClmIVX6fyIY/edit

@SarfrazS  Can you please check once by doing fresh installation that nothing is breaking when no records are there in db.

SarfrazS commented:
I not checked with a fresh installation. I checked load time with billing DB, Avinash shared with me.

ashu555 commented:
> If we give app name is agora billing, then Not able to install the development branch as well as this branch also.
> ![Annotation 2020-04-14 123237](https://user-images.githubusercontent.com/18435376/79196189-341b3000-7e4d-11ea-8d23-b42ea404b9cd.png)

Raise a seperate issue and assign it to me

avinash403 commented:
> Today i again tested with fresh installation nothing is breaking all pages are opening.
> Also tested with existing DB all data are showing on the dashboard.
> Dashboard performance(Load time) is better than development and live billing portal.
> The timezone issue is there but Ashutosh told us he fixed on Register PR.
> https://docs.google.com/document/d/10rh-LPE3HegoSMf5rAO1GyYzxCLRu04vClmIVX6fyIY/edit

Timezone issue I have fixed for dashboard @SarfrazS please retest.
@ashu555 I have added a common function to handle timezone conversion in `helpers.php`. Please have a look and try to use same function in new code that you will write

ashu555 commented on app/Http/helpers.php:
```diff
@@ -71,3 +71,22 @@ function successResponse($message = '', $data = '', $statusCode = 200)
+    // caching for 4 seconds so for consecutive queries, it will be readily available. And even if someone updates their
+    // timezone, it will start showing the new timezone after 4 seconds
+    $timezone = Cache::remember('timezone_'.Auth::user()->id, 5, function(){
+        return isset(Auth::user()->timezone) ? Auth::user()->timezone->name : 'UTC';
```
Do we need this check here? We can directly get the timezone name since the timezone saved is never null

avinash403 replied:
Just for the fallback. I wasn't confident that it can't be null, so added. Are we confident?

ashu555 replied:
Yes, even in the database directly, it cannot be set as null, unless ofcourse someone forcibly tries to do it.

avinash403 replied:
removed the check

ashu555 commented on app/Http/helpers.php:
```diff
@@ -71,3 +71,22 @@ function successResponse($message = '', $data = '', $statusCode = 200)
+    });
+
+    return ((new DateTime($dateTimeString))->setTimezone(new DateTimeZone($timezone)))->format($format);
+}
```
What happens if the system $timezone parameter passed in the DateTimeZone object does not matches PHP's list of timezones. For example-  we pass Asia/Calcutta from our system and DateTimeZone object  is expecting Asia/Kolkata as parameter.

avinash403 replied:
timezone is coming from database. Do we have such timezone in the DB?

avinash403 replied:
those cases must result in an internal server error else it will go unnoticed. We can decide that if we face such situations
